[compatibility version] Avoid resetting config when adding flags

yongruilin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:

Avoid resetting config of emulation verison and featuregates when adding flags
    This change introduces improvements to the component compatibility registry:

- Modify the kube-scheduler test server to create a separate ComponentGlobalsRegistry
- Update the compatibility registry to handle multiple flag configurations
- Enhance test cases to support emulation version mapping between components

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/129994

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md
```

siyuanfoundation commented on test/integration/examples/apiserver_test.go:
```diff
@@ -699,11 +699,12 @@ func prepareAggregatedWardleAPIServer(ctx context.Context, t *testing.T, namespa
-	componentGlobalsRegistry := testServer.ServerOpts.Options.GenericServerRunOptions.ComponentGlobalsRegistry
+	// Create a new registry since the testServer's ComponentGlobalsRegistry is already Set(),
+	// and wardle server would try to Set() again in the test.
+	componentGlobalsRegistry := basecompatibility.NewComponentGlobalsRegistry()
```
why creating a new instance? `testServer` already uses `NewComponentGlobalsRegistry`
https://github.com/kubernetes/kubernetes/blob/77d7f6380071e5719345aa34461c3e4619e0efab/cmd/kube-apiserver/app/testing/testserver.go#L198

yongruilin replied:
There is a comment I added to explain. Because wardleServer would call Set() in its command again, while testServer also calls Set() when it starts.

yongruilin replied:
Technically apiserver binary and wardle server binary would run separately outside of tests. So I think it is reasonable to separate the ComponentGlobalsResgistry?

yongruilin replied:
The mapping logic is handled in the command setup https://github.com/kubernetes/kubernetes/blob/f90682bf42f747f98a99b9e5b0ae415b2504c338/staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go#L163
So I update to register the "kube" component to the new registry with the same version and featuregate used in the `kastesting.StartTestServerOrDie` . Does it look right to you?

siyuanfoundation replied:
kube emulation version is set from mapping wardle emulation. So we need to set `--emulated-version=wardle=x.x` in the `extraKASFlags`

yongruilin replied:
So the updated kube emulation version is needed for starting the kube apiserver? I think this logic was implemented before right? As testServer has already started when https://github.com/kubernetes/kubernetes/blob/f90682bf42f747f98a99b9e5b0ae415b2504c338/staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go#L163 configured in the `NewCommandStartWardleServer()`

siyuanfoundation replied:
The function is set up. But the kube apiserver is started before the wardle server. If you do not pass in the wardle emulation version when starting the kube apiserver, the emulation version would just be equal to the binary version.

yongruilin replied:
Updated to add `--emulated-version=kube=%s` to testserver from `wardleEmulationVersion` .

BenTheElder commented:
/triage accepted

siyuanfoundation commented on cmd/kube-scheduler/app/testing/testserver.go:
```diff
@@ -98,6 +99,8 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
 
 	opts := options.NewOptions()
+	// Create an independent registry to avoid sharing state between different tests.
+	opts.ComponentGlobalsRegistry = compatibility.NewComponentGlobalsRegistry()
```
need to do something similar to https://github.com/kubernetes/kubernetes/blob/9a2a278a20d8b10a34809097a9a5e4e62a9cd9ef/cmd/kube-apiserver/app/testing/testserver.go#L194-L203 to pass the feature gate set in tests because most places uses the global `DefaultFeatureGate`

yongruilin replied:
Updated to the similar by registring the DefaultFeatureGate. I would have a separate PR to create a util func and use it in components.

siyuanfoundation commented on cmd/kube-controller-manager/app/options/options_test.go:
```diff
@@ -1480,6 +1481,9 @@ func TestWatchListClientFlagUsage(t *testing.T) {
 
 func TestWatchListClientFlagChange(t *testing.T) {
+	// Reset the global registry before the test, as DefaultComponentGlobalsRegistry is used by multiple tests.
+	compatibility.DefaultComponentGlobalsRegistry.Reset()
```
There might be a race condition if you depend on `Reset()` in tests. Better to use a `NewComponentGlobalsRegistry` and pass the feature gates through `SetFeatureGateDuringTest`

yongruilin replied:
Updated with a new commit. I think there is a problem with this test after this specific feature promoted to Beta.

siyuanfoundation commented on test/integration/examples/apiserver_test.go:
```diff
@@ -699,7 +714,14 @@ func prepareAggregatedWardleAPIServer(ctx context.Context, t *testing.T, namespa
+	componentGlobalsRegistry := basecompatibility.NewComponentGlobalsRegistry()
+	_, _ = componentGlobalsRegistry.ComponentGlobalsOrRegister(
+		basecompatibility.DefaultKubeComponent,
+		testServer.ServerOpts.Options.GenericServerRunOptions.ComponentGlobalsRegistry.EffectiveVersionFor(basecompatibility.DefaultKubeComponent).(basecompatibility.MutableEffectiveVersion),
+		testServer.ServerOpts.Options.GenericServerRunOptions.ComponentGlobalsRegistry.FeatureGateFor(basecompatibility.DefaultKubeComponent).(featuregate.MutableVersionedFeatureGate),
```
could just use brand new featuregate and effective version here. 
```
featureGate := utilfeature.DefaultMutableFeatureGate.DeepCopy()
effectiveVersion := utilcompatibility.DefaultKubeEffectiveVersionForTest()
```

yongruilin replied:
Updated.

siyuanfoundation commented on test/integration/examples/apiserver_test.go:
```diff
@@ -393,19 +400,27 @@ func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	// set the emulation version for the kube-apiserver testserver by mapping
+	// the wardle version to the kube version.
+	wardleEmulationVersion := version.MustParse(wardleEmulationVersionRaw)
+	kubeEmulationVersion := sampleserver.WardleVersionToKubeVersion(wardleEmulationVersion)
```
looks good!

siyuanfoundation commented on cmd/kube-controller-manager/app/options/options_test.go:
```diff
@@ -1493,7 +1496,9 @@ func TestWatchListClientFlagChange(t *testing.T) {
-	args := []string{fmt.Sprintf("--feature-gates=%v=true", clientgofeaturegate.WatchListClient)}
+	// To avoid race condition and multiple Set() calls on the registry,
+	// we need to pass a new registry and set the feature gate manually
+	args := []string{fmt.Sprintf("--feature-gates=%v=false", clientgofeaturegate.WatchListClient)}
```
How about add a new `func ReplaceFeatureGatesForTest(newFeatureGates Gates)` and use lock to bind client featuregate to the local feature gate, and restore to the original feature gate at the end of test?
Similar how it is done in `detectParallelOverrideEmulationVersion`

yongruilin replied:
Changed back to original test logic, probably need to improve it in another PR.
I just Deepcopy() the default featuregate, which should be added the client-go features, to the NewComponentGlobalRegistry. It should avoid the race condition and satisfy the original test purpose.

siyuanfoundation commented on staging/src/k8s.io/component-base/compatibility/registry.go:
```diff
@@ -324,6 +319,9 @@ func (r *componentGlobalsRegistry) SetFallback() error {
 	r.mutex.Lock()
 	defer r.mutex.Unlock()
+	if r.set {
+		return fmt.Errorf("componentGlobalsRegistry.Set() already called")
```
Based on https://github.com/kubernetes/enhancements/issues/4330#issuecomment-2687171933, it seems we should allow multiple calls of `Set()`. Can we return an Error only if there is a conflict of emulation version, or setting a feature gate to conflicting values?

yongruilin replied:
Then loose the restriction on multiple calls make sense.
> Can we return an Error only if there is a conflict of emulation version, or setting a feature gate to conflicting values?

The `Set()` just follows the `emulationVersionConfig` and `featureGatesConfig`, so what if these 2 actually changed on purpose?  "conflicting" is a bit vague to define, I think we should focus on the usage of the global var `DefaultGlobalComponentRegistry` vs `NewComponentRegistry()`, and let the `Set()` keep following the config derived from the flags.

siyuanfoundation replied:
if the intention is actually changing on purpose, then the code should call registry.Reset() intentionally. 
I agree the `DefaultGlobalComponentRegistry` global variable is an issue. We can followup on that. 
In the short term, allowing multiple calls of Set() and have some loose checks could unblock the issue of setting multiple components. WDYT?

siyuanfoundation commented on staging/src/k8s.io/component-base/compatibility/registry.go:
```diff
@@ -324,11 +320,52 @@ func (r *componentGlobalsRegistry) SetFallback() error {
+					if err != nil {
+						return fmt.Errorf("invalid value of %s=%s, err: %v", featureName, v, err)
+					}
+					if boolValue != globals.featureGate.Enabled(featuregate.Feature(featureName)) {
```
one tricky part is you would not be able to get the value of  features like `AllBeta` through the `featureGate.Enabled`.  Since the `featureGatesConfig` is a map of `componentName : featureValues`, and new flag values are just appended to `featureValues` each time `flag.Parse()` is called. I would suggest do something like:
```
for comp, fg := range r.featureGatesConfig {
		if err := validateNoConflictingFlagValue(fg) {
			return fmt.Errorf("setting conflicting feature value for component %s, err=%w", comp, err)
		}
}

func validateNoConflictingFlagValue(featureValues []string) error {
	featureToValues =: make(map[string]string)
	for _, featureValue : range featureValues {
		arr := strings.SplitN(featureValue, "=", 2)
		if len(arr) != 2 {
			return fmt.Errorf("malformed pair, expect string=string")
		}
		feature := strings.TrimSpace(arr[0])
		value := strings.TrimSpace(arr[1])
		if _, ok := featureToValues[feature]; ok {
			if featureToValues[feature] != value {
				return fmt.Errorf("setting conflicting feature value for feature %s", feature)
			}
		} else {
			featureToValues[feature] = value
		}
	}
	return nil
}
```

yongruilin replied:
The flagSet are new for each command, so the flag values are overridden instead of being appended.
https://github.com/kubernetes/component-base/blob/6d4c31484aea5d48318953956cb33314d08e9b3e/cli/flag/map_string_bool.go#L58-L62

yongruilin replied:
I would open a separate PR for resolving the conflict if needed. Adding the pointer doesn't seem to solve the problem for parsing? What I was thinking is to save the set config in the registry called `internalFeatureGateConfig`, which represent the one that has already been used to Set(). Then we can compare the `featureGateConfig` and the `internalFeatureGateConfig` to check the conflict.

siyuanfoundation commented on staging/src/k8s.io/component-base/compatibility/registry_test.go:
```diff
@@ -417,3 +417,88 @@ func assertVersionEqualTo(t *testing.T, ver *version.Version, expectedVer string
 	t.Errorf("expected: %s, got %s", expectedVer, ver.String())
 }
+
+func TestConfigConflicts(t *testing.T) {
```
please also add test case for `AllBeta` feature

siyuanfoundation commented on staging/src/k8s.io/component-base/compatibility/registry_test.go:
```diff
@@ -417,3 +417,88 @@ func assertVersionEqualTo(t *testing.T, ver *version.Version, expectedVer string
 	t.Errorf("expected: %s, got %s", expectedVer, ver.String())
 }
+
+func TestConfigConflicts(t *testing.T) {
```
can you make this test to use flags? to be as close as possible to real use case

yongruilin replied:
Outdated.

siyuanfoundation commented on staging/src/k8s.io/component-base/compatibility/registry.go:
```diff
@@ -324,11 +319,45 @@ func (r *componentGlobalsRegistry) SetFallback() error {
 	if err != nil {
 		return err
 	}
+	if r.set {
```
as we discussed offline, we probably do not need this validation any more because flag override was allowed pre-compat version. Can you clean up the code? Most of the other test changes are probably not needed anymore.

yongruilin replied:
Sounds good. I removed some change but keep some test related change which I think is better to keep to improve registry setup.

yongruilin committed:
chore: avoid resetting config of emulation verison and featuregates when adding flags

This change introduces improvements to the component compatibility registry:
- Modify the kube-scheduler test server to create a separate ComponentGlobalsRegistry
- Update the compatibility registry to handle multiple flag configurations
- Enhance test cases to support emulation version mapping between components

siyuanfoundation commented on cmd/kube-scheduler/app/testing/testserver.go:
```diff
@@ -98,6 +100,16 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
+	if err := componentGlobalsRegistry.Register(compatibility.DefaultKubeComponent, effectiveVersion, featureGate); err != nil {
+		return result, err
+	}
+	opts.ComponentGlobalsRegistry = componentGlobalsRegistry
```
need to add similar lines as https://github.com/kubernetes/kubernetes/blob/3a80473cda2da1c02ab0bf42bd3409ad6a54e6b1/cmd/kube-apiserver/app/testing/testserver.go#L373-L380

yongruilin replied:
Updated.

siyuanfoundation commented on staging/src/k8s.io/component-base/compatibility/registry.go:
```diff
@@ -226,11 +226,6 @@ func (r *componentGlobalsRegistry) AddFlags(fs *pflag.FlagSet) {
 			globals.featureGate.Close()
 		}
 	}
-	if r.emulationVersionConfig != nil || r.featureGatesConfig != nil {
```
still would be good to add a unit test where `AddFlags()` is called twice.

yongruilin replied:
Modify the TestFlags test to check on this.

siyuanfoundation commented on cmd/kube-scheduler/app/testing/testserver.go:
```diff
@@ -98,6 +100,28 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
+	featureGate := utilfeature.DefaultMutableFeatureGate.DeepCopy()
+	effectiveVersion := utilcompatibility.DefaultKubeEffectiveVersionForTest()
+	effectiveVersion.SetEmulationVersion(featureGate.EmulationVersion())
+	if !featureGate.EmulationVersion().EqualTo(utilfeature.DefaultMutableFeatureGate.EmulationVersion()) {
```
These need to be moved to after `componentGlobalsRegistry.Set()`

yongruilin commented:
/retest

siyuanfoundation commented:
/lgtm
Thanks for fixing this @yongruilin !

yongruilin commented:
/cc @liggitt 
Could you take a look as an approver?

haosdent commented:
> Avoid restting

Should be "resetting"?

BenTheElder reviewed (commented):
test/ changes look good, the rest looks good too but we should get OWNER input

BenTheElder commented:
/cc @kerthcet @smarterclayton

kerthcet commented on cmd/kube-scheduler/app/testing/testserver.go:
```diff
@@ -98,6 +100,16 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
 	fs := pflag.NewFlagSet("test", pflag.PanicOnError)
 
 	opts := options.NewOptions()
+	// set up new instance of ComponentGlobalsRegistry instead of using the DefaultComponentGlobalsRegistry to avoid contention in parallel tests.
```
Could you help me to understand when the data race will happen, I think the set() function below is protected by the lock.

kerthcet replied:
Shouldn't we avoid the conflict at the DefaultComponentGlobalRegistry by providing lock-protected functions, rather than copy-then-reset the configurations.

yongruilin replied:
If so, kube-apiserver and kube-scheduler will still be using the same global DefaultComponentGlobalRegistry when they running together in the test. Even we lock when `Set()`, they could still override each other blindly. One possible solution is to separate everything by components, but that requires a much larger scope of refactoring, and may not worth the effort. https://github.com/kubernetes/enhancements/issues/4330#issuecomment-2697813834

kerthcet replied:
Thanks @yongruilin got the point here. so it's not about the conflict but the overwritten problem, the same as k3s. 
/approve

yongruilin commented on cmd/kube-scheduler/app/testing/testserver.go:
```diff
@@ -106,6 +118,18 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
 	if err := opts.ComponentGlobalsRegistry.Set(); err != nil {
 		return result, err
 	}
+	if !featureGate.EmulationVersion().EqualTo(utilfeature.DefaultMutableFeatureGate.EmulationVersion()) {
```
Oh..it should be setting the defaultFeatureGate instead. I added the comment.
Basically it follows the logic in kube-apiserver https://github.com/kubernetes/kubernetes/blob/3a80473cda2da1c02ab0bf42bd3409ad6a54e6b1/cmd/kube-apiserver/app/testing/testserver.go#L370-L381

yongruilin commented on cmd/kube-scheduler/app/testing/testserver.go:
```diff
@@ -106,6 +118,18 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
+	}
+	for f := range utilfeature.DefaultMutableFeatureGate.GetAll() {
+		if featureGate.Enabled(f) != utilfeature.DefaultFeatureGate.Enabled(f) {
+			if err := utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%s=%v", f, featureGate.Enabled(f))); err != nil {
```
Yes

kerthcet reviewed (commented):
Just one question.

siyuanfoundation commented:
/cc @jpbetz for api-machinery review

jpbetz commented:
/approve
Relying on review from @siyuanfoundation and others

BenTheElder commented:
/approve

yongruilin commented:
/kind cleanup

pacoxu commented:
/test pull-kubernetes-e2e-gce

yongruilin commented:
/test pull-kubernetes-unit

serathius commented:
Looks like Tide is stuck on merging this PR https://prow.k8s.io/tide-history?repo=kubernetes%2Fkubernetes&branch=master

BenTheElder commented:
/hold
I think the PR needs a rebase but github isn't reflecting it until merge is attempted?
```json
{
  "insertId": "hvliggos8xhtifrc",
  "jsonPayload": {
    "sha": "50c543d99d8edb420780e7183a9427da1f4df30e",
    "func": "sigs.k8s.io/prow/pkg/tide.(*GitHubProvider).mergePRs",
    "branch": "master",
    "file": "sigs.k8s.io/prow/pkg/tide/github.go:291",
    "repo": "kubernetes",
    "merge-targets": [
      130079
    ],
    "pr": 130079,
    "msg": "Merge failed.",
    "base-sha": "73f54b67b29d77601b0bd42ad8b4992925b9df47",
    "component": "tide",
    "error": "PR was modified: Head branch is out of date. Review and try the merge again.",
    "controller": "sync",
    "org": "kubernetes",
    "level": "debug"
  },
  "resource": {
    "type": "k8s_container",
    "labels": {
      "pod_name": "tide-5b9cc8679d-d285s",
      "project_id": "k8s-infra-prow",
      "cluster_name": "prow",
      "namespace_name": "default",
      "container_name": "tide",
      "location": "us-central1"
    }
  },
  "timestamp": "2025-03-19T15:10:36Z",
  "severity": "DEBUG",
  "labels": {
    "logging.gke.io/top_level_controller_name": "tide",
    "k8s-pod/pod-template-hash": "5b9cc8679d",
    "compute.googleapis.com/resource_name": "gke-prow-prod-v1-30b07f94-padh",
    "logging.gke.io/top_level_controller_type": "Deployment",
    "k8s-pod/app": "tide"
  },
  "logName": "projects/k8s-infra-prow/logs/stderr",
  "receiveTimestamp": "2025-03-19T15:10:38.965003416Z"
}
```

BenTheElder commented:
/hold cancel
/lgtm
